### PR TITLE
I2C: clean up and remove stop condition from the middle of `write_read`

### PR DESCRIPTION
**Pull request by bugadani**
@bjoernQ I hope you don't mind me messing with your latest changes.

### CODE REVIEW

File: esp-hal/src/i2c.rs [hunk 1]
```diff
@@ -305,15 +344,14 @@ where
     ///   to indicate writing
     /// - `SR` = repeated start condition
     /// - `SP` = stop condition
-    pub fn transaction(
+    pub fn transaction<'a>(
         &mut self,
         address: u8,
-        operations: &mut [impl I2cOperation],
+        operations: impl Iterator<Item = Operation<'a>>,
```
Comment (Dominaezzz) [inline esp-hal/src/i2c.rs:350]: `IntoIterator`? (I think this is needed for slices maybe)
```suggestion
        operations: impl IntoIterator<Item = Operation<'a>>,
```

**Comment by bjoernQ** [comment]
> @bjoernQ I hope you don't mind me messing with your latest changes.

Not at all 😆 it's almost what I tried before - don't even sure what I did wrong before regarding the lifetime problems (maybe because of "Friday" 🤷‍♂️ )

Nice!

**Review by bjoernQ — APPROVED** [review]
Nice

File: esp-hal/src/i2c.rs (hunk 1 above)
Comment (bugadani) [reply esp-hal/src/i2c.rs:350]: Fair, I'll even add a test for this just to see.

Comment (bugadani) [reply esp-hal/src/i2c.rs:367]: Almost, but slices convert into an Iterator that returns references, which is annoying :)

**Comment by bugadani** [comment]
Sorry, much has changed since the last review. I believe the `write_read` changes are justified, although I haven't looked at waveforms myself.

**Comment by bjoernQ** [comment]
😢 yeah `write_read` wasn't a masterpiece 😢

**Review by jessebraham — APPROVED** [review]
LGTM

**Pull request merged**